Spam2: system for flagging nodes and comments to moderators and User moderation

keshavsethi opened the pull request:
I have added the feature of Flag Nodes in spam2 in which any user can mark node as a flag. It will not affect the status of the node.

Flag will just increase the flag count by 1. If flag is 0 then it means that the node is not flagged and it is ok. All the flagged nodes will be listed on the flags page of spam2. Right now I have done it for nodes and working for flag comments as well.(refer https://github.com/publiclab/plots2/pull/8107#issuecomment-653942167)

If  flag count is above 10 then flag icon also becomes red. 

Here are a few screenshots

![](https://user-images.githubusercontent.com/36025262/87270943-532e6a00-c4ef-11ea-88c1-d7ded81b4f11.gif)

![Screenshot from 2020-07-02 20-20-22](https://user-images.githubusercontent.com/36025262/86374270-ff778180-bca1-11ea-9479-927701f6af01.png)
![Screenshot from 2020-07-02 20-21-03](https://user-images.githubusercontent.com/36025262/86374281-01d9db80-bca2-11ea-96c1-68a939795fae.png)

Here flag count in shown in the status column and red icon indicates that the flag count is more than 10. Spamming that node will also unflag it as it is moderated.
![Screenshot from 2020-07-02 20-21-05](https://user-images.githubusercontent.com/36025262/86374299-043c3580-bca2-11ea-92a7-4121a4fae904.png)
 
@jywarren @cesswairimu @VladimirMikulic @ananyaarun @emilyashley @pydevsg @ebarry @Uzay-G Please review
Thanks!!

ebarry commented:
This is a great idea!
Who sees the red flag, and where? 

- Only moderators looking at the /spam dashboard? 
- Regular users browsing content?

keshavsethi commented:
> This is a great idea!
> Who sees the red flag, and where?
> 
> * Only moderators looking at the /spam dashboard?
> * Regular users browsing content?

Regular users can just flag a node. List of flag nodes and that red flag icon is only visible to moderators at /spam2/flags. 

Thanks!!

keshavsethi commented:
> Since everything is centred, you could as well centre this section with control buttons:
> https://i.imgur.com/YC7Xn5y.png

@VladimirMikulic Sure, I will do that. Thanks!!

pydevsg commented:
@keshavsethi , the UI is pretty good, have you made this response as well or it's currently made for the Desktop version?

keshavsethi commented:
@jywarren @cesswairimu @VladimirMikulic @ananyaarun @emilyashley @pydevsg @ebarry @Uzay-G 
I have also done flag comments in this PR. Here, any user can mark comment as a flag(it will not affect the status of a comment). Moderators can see those flag comments in /spam2/comments that have all the features like publish,spam, etc. flag is also shown and moderator can identify flag comments through the icon.
It is sorted as per flag count. I am currently working on its bulk moderation and tests
![Screenshot_2020-07-06 🎈 Public Lab _spam_comments(3)](https://user-images.githubusercontent.com/36025262/86542563-4aa8c300-bf34-11ea-8a4b-c7bbf2f88bac.png)
![Screenshot_2020-07-06 🎈 Public Lab _spam_comments](https://user-images.githubusercontent.com/36025262/86542597-7d52bb80-bf34-11ea-8918-113b623a194e.png)
@jywarren @cesswairimu @VladimirMikulic @ananyaarun @emilyashley @pydevsg @ebarry @Uzay-G 
Please review
Thanks!!

keshavsethi commented:
> @keshavsethi , the UI is pretty good, have you made this response as well or it's currently made for the Desktop version?

Yes, all the pages in /spam2 are responsive
![Screenshot_2020-07-06 🎈 Public Lab _spam_flags](https://user-images.githubusercontent.com/36025262/86542741-f0106680-bf35-11ea-8c2e-15bbe8af1055.png)

keshavsethi commented:
> @keshav234156 can you please share screenshot of the change, thanks

@VladimirMikulic @pydevsg I have aligned bulk nav 
![Screenshot from 2020-07-06 03-12-05](https://user-images.githubusercontent.com/36025262/86542800-a07e6a80-bf36-11ea-9f8d-b222ed7392d9.png)
Thanks!!

pydevsg commented:
Ah great work man , this UI is really clean and crisp. Keep up the good work @keshavsethi :tada:

jywarren commented:
Looking really nice! One suggestion is to use lots of Bootstrap tooltips:

https://getbootstrap.com/docs/4.4/components/tooltips/

to add short explanatory texts over things like the flag icon, without needing a lot of space! What do you think?

keshavsethi commented on app/controllers/spam2_controller.rb:
```diff
@@ -135,4 +149,36 @@ def batch_unban
+    if @node.flag == 0
+      flash[:notice] = 'Node already unflagged.'
+    else
+      @node.unflag_node
```
I have added an if statement to check only moderators and admins can unflag item 
Thanks

keshavsethi commented on app/models/comment.rb:
```diff
@@ -193,6 +193,18 @@ def publish
+    self
+  end
+
+  def unflag_comment
```
But for this, we have to show flag count to users as well. Right now unflag means only moderators can unflag (i.e reduce flag count to 0). Other users can just flag nodes i.e increase its flag count by 1

cesswairimu replied:
gotcha, cool, sounds good...thanks for the clarification

keshavsethi commented:
> Looking really nice! One suggestion is to use lots of Bootstrap tooltips:
> 
> https://getbootstrap.com/docs/4.4/components/tooltips/
> 
> to add short explanatory texts over things like the flag icon, without needing a lot of space! What do you think?

@jywarren I have added a few in bulk moderation nav.

keshavsethi commented:
> Hi @keshavsethi I would also like to request that you add some tests -- to demonstrate how it is supposed to work, and to ensure it's protected against future changes. I know this is a lot, but it is really a critical part of a big feature like this so I appreciate it!
> 
> * unit tests for changes to the models
> * functional tests for controller actions
> * system tests for javascript actions (pressing buttons, complex integrated sequences like several actions chained together)
> 
> Does this make sense? Thank you so much!

@jywarren Sure, I am working on tests. Should I do add them to the same PR?

keshavsethi commented:
@jywarren @cesswairimu can we merge this? 
Thanks!!

cesswairimu commented:
Hey @keshavsethi this is coming on great :tada: :tada: . If the tests are ready you can add them here then we can get this merged. Thanks

keshavsethi commented:
> Hey @keshavsethi this is coming on great   . If the tests are ready you can add them here then we can get this merged. Thanks

@cesswairimu @jywarren @pydevsg @VladimirMikulic @ebarry Tests are added for flag posts. Please review.
Thanks!!

keshavsethi commented:
I have updated the UI and have added alerts after any moderation. Please refer to the following Screenshot
![ezgif com-crop (5)](https://user-images.githubusercontent.com/36025262/87270943-532e6a00-c4ef-11ea-88c1-d7ded81b4f11.gif)
![Screenshot from 2020-07-13 09-30-25](https://user-images.githubusercontent.com/36025262/87270457-ca62fe80-c4ed-11ea-821c-eb162b3bf1ee.png)
@jywarren @cesswairimu @pydevsg Please review, Thanks!!

cesswairimu commented:
restarting travis

cesswairimu commented on test/functional/spam2_controller_test.rb:
```diff
@@ -228,4 +249,70 @@ def teardown
         assert_redirected_to "/dashboard"
     end
+    #flag Nodes
+    test 'Anyone user should be able to flag a note' do
```
Great job writing the tests...lets fix the grammar here 'Any user  should...' or 'Anyone should...'

keshavsethi replied:
Sorrry :sweat_smile:  for that, Done. Thanks!!

cesswairimu commented on test/functional/spam2_controller_test.rb:
```diff
@@ -228,4 +249,70 @@ def teardown
+        assert_equal 0, node.flag
+    end
+    #flag Comments
+    test 'Anyone user should be able to flag a comments' do
```
Same here  'Any user  should...' or 'Anyone should...'

keshavsethi replied:
Done Thanks!!

jywarren commented:
Hi @keshavsethi this is looking tremendous. I wonder if "Flag this node" should be "Flag this to moderators" as that way it's known who sees the flags, and the vocab word "node" isn't as confusing to folks? 

Also, i wanted to see if you had any thoughts on how to complete #8122 because it occurred to me that it would make reviewing big PRs like this much easier; anyone could just boot it up in GitPod right from the PR! @Tlazypanda this could benefit some of your work as well. Would either of you be interested in helping to make progress on #8122 and that could speed up our review time quite a bit! 🎉

jywarren commented on test/functional/spam2_controller_test.rb:
```diff
@@ -228,4 +249,70 @@ def teardown
+        post :remove_flag_comment, params: { id: comment.id }
+        comment = assigns(:comment)
+        assert_equal 0, comments(:second).flag
+    end
```
These are great -- excellent for underlying functionality!

jywarren commented on test/system/spam2_test.rb:
```diff
@@ -33,6 +33,13 @@ def setup
     page.assert_selector("div.alert", text: "Content spammed")
   end
 
+  test "unflag post in spam2" do
```
This is great too! Do you think it'd be possible to add a couple for regular users flagging posts, and for some other specific review action by moderators?

keshavsethi replied:
@jywarren I have added a system test where user can mark node as a flag. Please review Thanks!!

keshavsethi commented:
> Hi @keshavsethi this is looking tremendous. I wonder if "Flag this node" should be "Flag this to moderators" as that way it's known who sees the flags, and the vocab word "node" isn't as confusing to folks?
> 
> Also, i wanted to see if you had any thoughts on how to complete #8122 because it occurred to me that it would make reviewing big PRs like this much easier; anyone could just boot it up in GitPod right from the PR! @Tlazypanda this could benefit some of your work as well. Would either of you be interested in helping to make progress on #8122 and that could speed up our review time quite a bit!

@jywarren That typo is done and  as #8122 is done Can we merge this. Thanks!!

keshavsethi commented:
@jywarren @cesswairimu can we merge this ??

cesswairimu commented:
Yeah looks good...we just need @jywarren final review against the changes he requested

keshavsethi commented:
@jywarren can you please review and merge this as my future PRs are dependent on this. Please review 
Thanks!!

keshavsethi commented:
@jywarren @cesswairimu you can test spam2 in gitpod Please review Thanks!!

keshavsethi commented:
@jywarren @cesswairimu @pydevsg I have added the User moderation page here where the moderator can moderate users. I have also added filters for admin moderators and banned users. Can you Please review and merge this PR as I am stuck here. This PR is getting bigger and has so many changes. You can review it in gitpod. Thanks!!
![Screenshot from 2020-07-20 10-52-26](https://user-images.githubusercontent.com/36025262/87902473-54274480-ca77-11ea-934b-18334c7ec82e.png)
In my future PRs, I am planning to add a search feature and **Queue** where the moderator can moderate only those nodes which have tags that they are following. 
Please review 
Thanks!!

keshavsethi commented:
@jywarren @cesswairimu I have updated the UI a bit and completed user bulk moderation. Please review!!
This PR is getting Bigger, Please review and merge if it is fine. :pray: 
![Screenshot from 2020-07-21 21-23-26](https://user-images.githubusercontent.com/36025262/88081609-3f919c00-cb9e-11ea-8b75-854faccfca9b.png)
Please review and merge if it is fine. Thanks!!

jywarren commented:
Reviewing this now! Thanks @keshavsethi for your patience. I hope we can work more modularly and that GitPod helps us review faster, so you get un-stuck!

jywarren commented:
It's booting up in GitPod! 🕐 😄

keshavsethi commented:
I have made some changes in gitpot https://github.com/publiclab/plots2/pull/8162 It is working fine there.

jywarren commented:
OK!

Nice!

![Screen Shot 2020-07-21 at 1 30 40 PM](https://user-images.githubusercontent.com/24359/88087102-6926d500-cb56-11ea-839c-359312b4a9d8.png)

jywarren commented:
Hmm. i'm still getting hit by `Mysql2::Error: Expression #3 of SELECT list is not in GROUP BY clause` so i'm going to try to resolve that first! But the page looks great so i am pretty confident it'll be good to merge today!

jywarren commented on test/system/dashboard_test.rb:
```diff
@@ -22,5 +22,16 @@ class DashboardTest < ApplicationSystemTestCase
   end
-
+  
+  test "User can flag a node from dashboard" do
```
Awesome. Thanks!

jywarren commented:
I'm content with how this looks so far! Thank you! The tests were super helpful and I loved being able to try it out on GitPod. Thanks and congratulations, @keshavsethi !!!

keshavsethi commented:
Thank you so much @jywarren  :tada:  :tada: :heart:
